src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/P11SecretKeyFactory.java

```diff
@@ -184,15 +323,137 @@ static P11Key convertKey(Token token, Key key, String algo,
         if (p11Key != null) {
             return p11Key;
         }
-        if (!"RAW".equalsIgnoreCase(key.getFormat())) {
-            throw new InvalidKeyException("Encoded format must be RAW");
+        if (key instanceof PBEKey pbeKey) {
+            ki = ki == null ? getKeyInfo(keyAlgo) : ki;
+            if (ki instanceof PBEKeyInfo pbeKi) {
+                try {
+                    p11Key = derivePBEKey(token, getPbeKeySpec(pbeKey), pbeKi);
+                } catch (InvalidKeySpecException e) {
+                    throw new InvalidKeyException(e);
+                }
+            } else {
+                throw new InvalidKeyException("Cannot derive unknown " +
+                        keyAlgo + " algorithm");
+            }
+        } else {
+            if (si instanceof PBEKeyInfo) {
+                throw new InvalidKeyException("PBE service requires a PBE key");
+            }
+            if (!"RAW".equalsIgnoreCase(key.getFormat())) {
+                throw new InvalidKeyException("Encoded format must be RAW");
+            }
+            byte[] encoded = key.getEncoded();
```

Comment:
Good

Reply:
We discussed this change with @franferrax and have some concerns. The method Key::getEncoded does not document that a copy will be returned, and this would change the current behavior and affect non-PBE cases. In practical terms, it would mean that a key directly or indirectly converted to a P11Key would be destroyed if it does not return a clone in its getEncoded method. We suggest to make the caller responsible and keep the existing behavior. I.e.: if we call with a SecretKeySpec —whose ::getEncoded returns a clone—, the caller will need to (optionally) clear the key. What do you think?